src/cargo/core/compiler/job_queue.rs

```diff
@@ -934,8 +933,76 @@ impl<'cfg> DrainState<'cfg> {
         let report_id = on_disk_reports.last_id();
 
         if bcx.build_config.future_incompat_report {
-            let rendered = on_disk_reports.get_report(report_id, bcx.config).unwrap();
-            drop(bcx.config.shell().print_ansi_stderr(rendered.as_bytes()));
+            let upstream_info = package_ids
+                .iter()
+                .map(|package_id| {
+                    let manifest = bcx.packages.get_one(*package_id).unwrap().manifest();
+                    format!(
+                        "
+  - {name}
+    - Repository: {url}
+    - Detailed warning command: `cargo report future-incompatibilities --id {id} --crate '{name}'",
+                        name = package_id,
```

Comment:
I'm a little concerned about using the `Display` of `PackageId`.  One problem is that it can be very verbose.  Another problem is that it can contain spaces.  Can this maybe use a minimized package ID spec?  Maybe something like `name:1.2.3`? There is code to deal with duplicate name/version pairs ([here](https://github.com/rust-lang/cargo/blob/a821e2cb24d7b6013433f069ab3bad53d160e100/src/cargo/core/package_id_spec.rs#L264-L276)), but it is not exposed and could be awkward to use here. So I'm thinking just ignore that possibility (it is exceedingly rare), and just use the name:version string.